File: A02/C++/P01/1/f.-CribaEratostenes.cpp
#include <iostream>
#include <iomanip>
#include <sstream>
#include <string>
#include <vector> 
#include <ctime>
#include <cmath>
using namespace std;

string formatTime(long double t) {
	ostringstream time;
	int hour, min, sec, milisec;
	
	hour = t/3600;
	t = t - (hour*3600);
	min = t/60;
	t = t - (min*60);
	sec = t;
	t = t - (sec/1);
	milisec = t*10000;

	time << setfill('0') <<
		setw(2) << hour << ':' <<
		setw(2) << min << ':' <<
		setw(2) << sec << '.' << 
		setw(4) << milisec;
	
	return time.str();
}

void criba(long long num, vector<bool> &primes){
	primes = vector<bool>(num, true);
	for(long long i = 2; i <= sqrt(num); i++) {
		for (long long j = i; j*i <= num; j++) {
			primes[(i*j)-1]=0;
		}
	}
}

bool isPrime(long long num, vector<bool> primes) {
	return primes[num-1];
}

int main() {
	vector<bool> primes;
	long long number;
	bool prime;
	clock_t t;
	string time;
	
	cin >> number;
	
	t = clock();
	criba(number, primes);
	prime = isPrime(number, primes);
	t = clock() - t;

	/*for(int i = 0; i < primes.size(); i++) 
		cout << primes.at(i) << " ";
	cout << endl;*/

	time = formatTime(((long double)t)/CLOCKS_PER_SEC);

	cout << number << ' ';
	prime? cout << "si" : cout << "no";
	cout << " es primo." << endl;
	cout << "Tiempo: " << time << endl;

	return 0;
}

File: A02/Python/P01/1/f.-CribaEratostenes.py
import time
import math

def format_time(t):
    time_formatted = ""
    hour = minute = second = milisecond = int(0)

    hour = int(t//3600);
    t = t - (hour*3600)
    minute = int(t//60)
    t = t - (minute*60)
    second = int(t//1)
    t = t - second
    t = int(t*10000)
    milisecond = t
    
    time_formatted += (("0" + str(hour)) if hour < 10 else str(hour)) + ':'
    time_formatted += (("0" + str(minute)) if minute < 10 else str(minute)) + ':'
    time_formatted += (("0" + str(second)) if second < 10 else str(second)) + '.'
    time_formatted += '{:0>4}'.format(milisecond)

    return time_formatted


def criba(num):
    primes = {key: True for key in range(num+1)}
    for i in range(2, int(math.sqrt(num))+1):
        if primes.get(i):
            j = i
            while (j*i < num+1):
                primes[(i*j)-1] = False;
                j += 1

    return primes


def is_prime(num, primes):

    return primes.get(num-1)


def main():
    number = 0
    primes = {}
    prime = True
    exec_time = ""
    t = 0.0

    number = int(input())
    
    primes = criba(number)
    t = time.time()
    prime = is_prime(number, primes)
    t = time.time() - t
    exec_time = format_time(t);

    print(number, "si" if prime else "no", "es primo")
    print("Total:", exec_time)

if __name__ == "__main__":
    main()
